### ISSUE TITLE
rclone serve s3: invalid error response when using wrong password

### ISSUE THREAD

**Issue by static-moonlight** [rclone/rclone#7749]
**rclone serve s3** can't handle invalid credentials very well:

1. Open a terminal and run `rclone server s3` (in my case, I use a container):
    ```bash
    $ podman run --rm --env=RCLONE_ADDR=:8080 --env=RCLONE_AUTH_KEY=admin,admin --publish=8080:8080 --volume=/tmp/data:/data docker.io/rclone/rclone:1.66 serve s3 /data
    ```
2. Open a 2nd terminal and run a simple request with rclone as a client (the password is wrong on purpose):
    ```bash
    $ RCLONE_S3_ENDPOINT=http://localhost:8080 RCLONE_S3_ACCESS_KEY_ID=admin RCLONE_S3_SECRET_ACCESS_KEY=admin2 ./rclone lsf -vv s3:
    ```
    `/home/tony/.config/rclone/rclone.conf`:
    ```
    [s3]
    type = s3
    provider = Other
    env_auth = true
    ```
4. Result: I can see many HTTP 403 responses with Wireshark and rclone seems to ignore all of them and doesn't do much for roughly a minute until it finally exists with an error:
    ```bash
	2024/04/09 22:07:58 DEBUG : Setting default for s3-access-key-id="admin" from environment variable RCLONE_S3_ACCESS_KEY_ID
	2024/04/09 22:07:58 DEBUG : Setting default for s3-secret-access-key="admin2" from environment variable RCLONE_S3_SECRET_ACCESS_KEY
	2024/04/09 22:07:58 DEBUG : Setting default for s3-endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : rclone: Version "v1.66.0" starting with parameters ["./rclone" "lsf" "-vv" "s3:"]
	2024/04/09 22:07:58 DEBUG : Creating backend with remote "s3:"
	2024/04/09 22:07:58 DEBUG : Using config file from "/home/tony/.config/rclone/rclone.conf"
	2024/04/09 22:07:58 DEBUG : Setting s3_access_key_id="admin" from environment variable RCLONE_S3_ACCESS_KEY_ID
	2024/04/09 22:07:58 DEBUG : Setting s3_secret_access_key="admin2" from environment variable RCLONE_S3_SECRET_ACCESS_KEY
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : s3: detected overridden config - adding "{u_K4u}" suffix to name
	2024/04/09 22:07:58 DEBUG : Setting s3_access_key_id="admin" from environment variable RCLONE_S3_ACCESS_KEY_ID
	2024/04/09 22:07:58 DEBUG : Setting s3_secret_access_key="admin2" from environment variable RCLONE_S3_SECRET_ACCESS_KEY
	2024/04/09 22:07:58 DEBUG : Setting s3_endpoint="http://localhost:8080" from environment variable RCLONE_S3_ENDPOINT
	2024/04/09 22:07:58 DEBUG : Resolving service "s3" region "us-east-1"
	2024/04/09 22:07:58 DEBUG : fs cache: renaming cache item "s3:" to be canonical "s3{u_K4u}:"
	2024/04/09 22:08:43 ERROR : : error listing: SerializationError: failed to unmarshal error message
		status code: 403, request id: , host id: 
	caused by: UnmarshalError: failed to unmarshal error message
		00000000  3c 3f 78 6d 6c 20 76 65  72 73 69 6f 6e 3d 22 31  |<?xml version="1|
	00000010  2e 30 22 20 65 6e 63 6f  64 69 6e 67 3d 22 55 54  |.0" encoding="UT|
	00000020  46 2d 38 22 3f 3e 0a 3c  65 72 72 6f 72 52 65 73  |F-8"?>.<errorRes|
	00000030  70 6f 6e 73 65 3e 3c 43  6f 64 65 3e 53 69 67 6e  |ponse><Code>Sign|
	00000040  61 74 75 72 65 44 6f 65  73 4e 6f 74 4d 61 74 63  |atureDoesNotMatc|
	00000050  68 3c 2f 43 6f 64 65 3e  3c 4d 65 73 73 61 67 65  |h</Code><Message|
	00000060  3e 54 68 65 20 72 65 71  75 65 73 74 20 73 69 67  |>The request sig|
	00000070  6e 61 74 75 72 65 20 77  65 20 63 61 6c 63 75 6c  |nature we calcul|
	00000080  61 74 65 64 20 64 6f 65  73 20 6e 6f 74 20 6d 61  |ated does not ma|
	00000090  74 63 68 20 74 68 65 20  73 69 67 6e 61 74 75 72  |tch the signatur|
	000000a0  65 20 79 6f 75 20 70 72  6f 76 69 64 65 64 2e 20  |e you provided. |
	000000b0  43 68 65 63 6b 20 79 6f  75 72 20 6b 65 79 20 61  |Check your key a|
	000000c0  6e 64 20 73 69 67 6e 69  6e 67 20 6d 65 74 68 6f  |nd signing metho|
	000000d0  64 2e 3c 2f 4d 65 73 73  61 67 65 3e 3c 2f 65 72  |d.</Message></er|
	000000e0  72 6f 72 52 65 73 70 6f  6e 73 65 3e              |rorResponse>|

	caused by: expected element type <Error> but have <errorResponse>
	2024/04/09 22:08:43 DEBUG : 6 go routines active
	2024/04/09 22:08:43 Failed to lsf with 2 errors: last error was: error in ListJSON: SerializationError: failed to unmarshal error message
		status code: 403, request id: , host id: 
	caused by: UnmarshalError: failed to unmarshal error message
		00000000  3c 3f 78 6d 6c 20 76 65  72 73 69 6f 6e 3d 22 31  |<?xml version="1|
	00000010  2e 30 22 20 65 6e 63 6f  64 69 6e 67 3d 22 55 54  |.0" encoding="UT|
	00000020  46 2d 38 22 3f 3e 0a 3c  65 72 72 6f 72 52 65 73  |F-8"?>.<errorRes|
	00000030  70 6f 6e 73 65 3e 3c 43  6f 64 65 3e 53 69 67 6e  |ponse><Code>Sign|
	00000040  61 74 75 72 65 44 6f 65  73 4e 6f 74 4d 61 74 63  |atureDoesNotMatc|
	00000050  68 3c 2f 43 6f 64 65 3e  3c 4d 65 73 73 61 67 65  |h</Code><Message|
	00000060  3e 54 68 65 20 72 65 71  75 65 73 74 20 73 69 67  |>The request sig|
	00000070  6e 61 74 75 72 65 20 77  65 20 63 61 6c 63 75 6c  |nature we calcul|
	00000080  61 74 65 64 20 64 6f 65  73 20 6e 6f 74 20 6d 61  |ated does not ma|
	00000090  74 63 68 20 74 68 65 20  73 69 67 6e 61 74 75 72  |tch the signatur|
	000000a0  65 20 79 6f 75 20 70 72  6f 76 69 64 65 64 2e 20  |e you provided. |
	000000b0  43 68 65 63 6b 20 79 6f  75 72 20 6b 65 79 20 61  |Check your key a|
	000000c0  6e 64 20 73 69 67 6e 69  6e 67 20 6d 65 74 68 6f  |nd signing metho|
	000000d0  64 2e 3c 2f 4d 65 73 73  61 67 65 3e 3c 2f 65 72  |d.</Message></er|
	000000e0  72 6f 72 52 65 73 70 6f  6e 73 65 3e              |rorResponse>|

	caused by: expected element type <Error> but have <errorResponse>
    ```

Used rclone version (client):
```bash
$ ./rclone --version
rclone v1.66.0
- os/version: debian 12.1 (64 bit)
- os/kernel: 6.1.0-18-amd64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.22.1
- go/linking: static
- go/tags: none
```

Since I use rclone a lot as S3 client and never had such problems, I suspect, that `rclone serve s3` might not be fully complient to the S3 specification (yet) and sends an invalid error response. According to [Amazon's S3 documentation](https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html), such responses should look like this:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Error>
  [...]
</Error>
```
... but apparently the response sent by `rclone serve s3`, looks like this:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<errorResponse>
  [...]
</errorResponse>
```
 I also suspect that the rclone S3 client doesn't expect such a response and tries again and again for about a minute until it finally fails. I wouldn't be surprised, if such invalid error responses might happen for other errors as well.

EDIT: I'm also not sure that this message is the best way to tell the client that the password is wrong:

"_The request signature we calculated does not match the signature you provided. Check your key and signing method._"

**Comment by ncw** [comment]
Hmm, interesting!

Can you check if rclone does the same thing with incorrect credentials with AWS S3?

I suspect this will need to be fixed in the upstream library.

**Comment by ncw** [comment]
Please give this a try

[v1.67.0-beta.7876.5c6c39510.fix-7749-s3-serve-error](https://beta.rclone.org/branch/fix-7749-s3-serve-error/v1.67.0-beta.7876.5c6c39510.fix-7749-s3-serve-error/) on branch [fix-7749-s3-serve-error](https://github.com/rclone/rclone/tree/fix-7749-s3-serve-error) (uploaded in 15-30 mins)

**Comment by ncw** [comment]
This fixes the XML for me

Before (the master branch)

```
2024/04/18 11:59:02 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2024/04/18 11:59:03 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2024/04/18 11:59:03 DEBUG : HTTP REQUEST (req 0xc000b88500)
2024/04/18 11:59:03 DEBUG : GET / HTTP/1.1
Host: localhost:8080
User-Agent: rclone/v1.67.0-beta.7896.2063e017d
Authorization: XXXX
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20240418T105903Z
Accept-Encoding: gzip

2024/04/18 11:59:03 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2024/04/18 11:59:03 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2024/04/18 11:59:03 DEBUG : HTTP RESPONSE (req 0xc000b88500)
2024/04/18 11:59:03 DEBUG : HTTP/1.1 403 Forbidden
Content-Length: 236
Content-Type: text/xml; charset=utf-8
Date: Thu, 18 Apr 2024 10:59:03 GMT

<?xml version="1.0" encoding="UTF-8"?>
<errorResponse><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message></errorResponse>
```

After (with the fix above)

```
2024/04/18 12:00:20 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2024/04/18 12:00:20 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2024/04/18 12:00:20 DEBUG : HTTP REQUEST (req 0xc000d10100)
2024/04/18 12:00:20 DEBUG : GET / HTTP/1.1
Host: localhost:8080
User-Agent: rclone/v1.67.0-beta.7876.5c6c39510.fix-7749-s3-serve-error
Authorization: XXXX
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20240418T110020Z
Accept-Encoding: gzip

2024/04/18 12:00:20 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2024/04/18 12:00:20 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2024/04/18 12:00:20 DEBUG : HTTP RESPONSE (req 0xc000d10100)
2024/04/18 12:00:20 DEBUG : HTTP/1.1 403 Forbidden
Content-Length: 236
Content-Type: text/xml; charset=utf-8
Date: Thu, 18 Apr 2024 11:00:20 GMT

<?xml version="1.0" encoding="UTF-8"?>
<ErrorResponse><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message></ErrorResponse>
```

I tested this with running this in one terminal

    rclone serve -vv --auth-key user,pass s3 /tmp/src/

And this in another (note the incorrect `secret_access_key`)

    rclone lsf -vv ':s3,provider=Rclone,access_key_id=user,secret_access_key=passX,endpoint="http://localhost:8080":' --dump bodies

What problem are you having and how can I replicate it? Preferably with terminal commands like the above not with containers.

**Comment by ncw** [comment]
Ah, sorry I mis-understood the problem.

Can you try this please

[v1.67.0-beta.7876.2dc6016de.fix-7749-s3-serve-error](https://beta.rclone.org/branch/fix-7749-s3-serve-error/v1.67.0-beta.7876.2dc6016de.fix-7749-s3-serve-error/) on branch [fix-7749-s3-serve-error](https://github.com/rclone/rclone/tree/fix-7749-s3-serve-error) (uploaded in 15-30 mins)

PS Incorrect root XML nodes are very common - I found a bug exactly like that in Google Cloud Storage's S3 implementation recently!

**Comment by ncw** [comment]
Thanks for testing. I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.67 

I'm going to ignore the `<RequestId>` and `<HostId>` elements for the moment as I think they are mostly useful for the cloud provider to debug stuff and since `rclone serve s3` doesn't have any error tracing or multiple hosts I think we can leave it out for the moment!
